Add client_key_api support

svoynow opened the pull request:
A simple implementation of the Erlang `:ssh_client_key_api` behavior, to
to make it easier to specify SSH keys and known_hosts files independently of
any particular users home directory

pmeinhardt commented on test/sshkit/client_key_api_test.exs:
```diff
@@ -0,0 +1,82 @@
+      known_hosts_file: known_hosts_file,
+      key_file: key_file
+    }
+  end
```
I am always a bit cautious about file-system operations - esp. `rm`s - on the developer/local machine…

Also, always using the same paths prevents these test from being run in parallel. From the top of my head: Could we use tempfiles instead?

svoynow replied:
tempfiles are the right call. I'll make it so.

pmeinhardt commented on lib/sshkit/ssh/client_key_api.ex:
```diff
@@ -0,0 +1,130 @@
+    |> has_fingerprint(key, hostname)
+  end
+
+  def user_key(_alg, opts) do
```
This seems to completely disregard the actual algorithm passed.

According to the docs, the `Algorithm` parameter is there for:

> Host key algorithm. Is to support `'ssh-rsa'| 'ssh-dss'` but more algorithms can be handled.

If these need different handling, it may be better to explicitly add `user_key/2` functions which match on the supported algos, e.g. `def user_key('ssh-rsa', options) do …`.

svoynow replied:
I'll see what needs to be done to support the other algorithms

pmeinhardt commented on lib/sshkit/ssh/client_key_api.ex:
```diff
@@ -0,0 +1,130 @@
+          Error: unknown fingerprint found for #{inspect hostname} #{inspect key}.
+          You either need to add a known good fingerprint to your known hosts file for this host,
+          *or* pass the accept_hosts option to your client key callback
+          """        
```
Whether we get here or not depends only on the value of `accept_hosts(opts)`. How would adding "a known good fingerprint" to the hosts file change the outcome here? The first half of the message sentence might be misleading (or I'm just confused)?

svoynow replied:
I believe that this callback never gets called if the `is_host_key` callback returns true.

svoynow commented:
1) As far as the `user_dir` option goes. That probably definitely works for most cases, but this seemed like it was the most flexible option, for example if you needed to perform multiple operations that used different keys,  rather than a single one.
2) I went back and forth on the IO devices vs. files question. On further consideration, I think you're right that passing IO devices and caching their contents is probably the way to go.
3) I did look into `ssh-agent` functionality a little bit. If it was going to be implemented, I think this would be the place to put it. I don't currently have a use-case for it myself.

svoynow commented:
Giving it a little more thought, one problem with the IO device approach is dealing with the case of accepting a new host fingerprint.
AFAIK, you can't reset a StringIO, so there's no simple way to make sure subsequent calls use the new data.
Store it in / read it from an `Agent` maybe?

svoynow commented:
@pmeinhardt, I think I've addressed the issues you raised.
It's currently using IO devices instead of paths, which has allowed me to test it without the use of files altogether. 
I'm also only accepting key algorithms that erlang seems to know how to decode.

svoynow commented:
@pmeinhardt, anything else I can do?

pmeinhardt commented:
Oh damn. Has it been that long… Sorry, I just need a to take some time this evening to go through it. All in all, I'm hoping this is something people'll use. For me personally, ssh-agent support would be better, but this looks like it can be an alternative, depending on your setup/workflow.

I'll set a reminder and from what I've seen, I think everything should be good. Thanks for the nudge again 👍

svoynow commented:
Our preference was for this just to be able to keep all of our configuration in Elixir-land, but I can see a use case for ssh-agent too. I did a brief

pmeinhardt commented:
I think part of your message got lost? "I did a brief…"

svoynow commented:
@pmeinhardt ping

pmeinhardt commented:
Hey @svoynow, I just looked over this again and something struck me:
Your `ClientKeyAPI` is actually usable *with or without* SSHKit.

In particular, it does not depend on anything in SSHKit and - at the same time - may be useful for other projects that work with Erlang's `:ssh_connection` and need custom key handling.

Therefore: **How about publishing this as an independent package?**

I can see the following advantages:

- People who need this behavior can then pull this in *even when they're not using SSHKit*
- People who are using SSHKit can *optionally* use it by passing it via the `:key_cb` option
- You have a smaller, decoupled standalone package that you can work on and release at your own pace

Let me know what you think.

Cheers 😃 

cc @brienw, @holetse

pmeinhardt commented:
…and seeing that the *same code* has [landed in SSHEx](https://github.com/rubencaro/sshex/blob/dc0913eaf4a46a9518bb21c0f2aca0e8147d6d90/lib/sshex/configurable_client_keys.ex) just asks for a separate package 😛 😁

svoynow commented:
Sounds good to me.

pmeinhardt commented:
@svoynow When you published it, let us know. Then we can have an `ADDONS.md` or similar here with links to modules usable in combination with SSHKit 😃
